platform/mds/mds/src/main/java/org/motechproject/mds/service/MdsOsgiBundleApplicationContextListener.java

```diff
@@ -0,0 +1,176 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MdsOsgiBundleApplicationContextListener.class);
+    private static final String WARN_MESSAGE= "%s failed to start but it was successfully restarted.";
+    private static final String CRITICAL_MESSAGE= "%s failed to start and the MDS module was unable to restart it.";
+
+    public static final String MESSAGE_SUBJECT = "org.motechproject.message";
+    public static final String MESSAGE_KEY = "message";
+    public static final String LEVEL_KEY = "level";
+    public static final String MODULE_NAME_KEY = "moduleName";
+    public static final String TIMEOUT_EXCEPTION_MESSAGE = "Application context initialization for '%s' has timed out waiting for";
+
+    private Set<String> restartedBundles = new HashSet<>();
+    private AllBundleFailsReports allBundleFailsReports;
+    private JdoTransactionManager transactionManager;
+    private OsgiEventProxy osgiEventProxy;
+    private SettingsService settingsService;
+
+    @Override
+    public void onOsgiApplicationEvent(OsgiBundleApplicationContextEvent event) {
+
+        if (!settingsService.getModuleSettings().isRestartModuleAfterTimeout()) {
+            return;
+        }
+
+        String symbolicName = event.getBundle().getSymbolicName();
+        if (event instanceof OsgiBundleContextFailedEvent) {
+            if (!restartedBundles.contains(symbolicName)) { // We want refresh bundles only one time
+                OsgiBundleContextFailedEvent failedEvent = (OsgiBundleContextFailedEvent) event;
+                final String failureCauseMsg = failedEvent.getFailureCause().getMessage();
+                Bundle bundle = failedEvent.getBundle();
+
+                // we will restart bundle only after timeout
+                // this is temporary solution until we don't fix problems with context
+                if (!failureCauseMsg.startsWith(String.format(TIMEOUT_EXCEPTION_MESSAGE, symbolicName))) {
+                    return;
+                }
+
+                LOGGER.error("Received context failed event {} from {}", event, symbolicName);
+                LOGGER.error("{} failed to start due to {}", symbolicName, failureCauseMsg);
+
+                logFailReport(symbolicName, failureCauseMsg, BundleRestartStatus.IN_PROGRESS);
+                restartedBundles.add(symbolicName);
+
+                LOGGER.info("Trying to restart {} bundle", symbolicName);
+                try {
+                    bundle.stop();
+                    bundle.start();
+                } catch (BundleException e) {
+                    LOGGER.error("Cannot restart {} bundle due to {}", symbolicName, e);
+                }
+            } else {
+                OsgiBundleContextFailedEvent failedEvent = (OsgiBundleContextFailedEvent) event;
+                LOGGER.error("Cannot restart {} bundle due to {}", symbolicName, failedEvent.getFailureCause().getMessage());
```

Comment:
The current message implies that we cannot restart it because of the failure cause, but in truth we cannot restart it since we already tried. Change to something along the lines

cannot restart {} bundle since it we already attempted it.